src/daft-table/src/lib.rs

```diff
@@ -459,12 +549,37 @@ impl Table {
             }
             seen.insert(name.clone());
         }
-        let schema = Schema::new(fields)?;
-        Table::new(schema, result_series)
+        let new_schema = Schema::new(fields)?;
+
+        let has_agg_expr = exprs.iter().any(|e| matches!(e.as_ref(), Expr::Agg(..)));
+        let num_rows = match (has_agg_expr, self.len()) {
+            // "Normal" case: the final cardinality is the max(*results_lens, self.len())
+            // This correctly accounts for broadcasting of literals, which can have unit length
+            (false, self_len) if self_len > 0 => result_series
+                .iter()
+                .map(|s| s.len())
+                .chain(std::iter::once(self.len()))
+                .max()
+                .unwrap(),
+            // "Empty" case: when no aggregation is applied, the expected result should also be empty
+            (false, _) => 0,
+            // "Aggregation" case: the final cardinality is the max(results' lens)
+            // We discard the original self.len() because we expect aggregations to change
+            // the final cardinality. Aggregations on empty tables are expected to produce unit length results.
+            (true, _) => result_series.iter().map(|s| s.len()).max().unwrap(),
+        };
```

Comment:
The logic for num_rows here was tricky to get correct. Reviewers should pay a little more attention to this block of code.

It was a little difficult to get the "correct" logic for how many rows would result from a call to eval_expression_list, which can contain cardinality-modifying expressions such as UDFs and aggregations.